### ISSUE TITLE
Fix regression in size_range() calculation

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#392]
The attempt to fix [STR 3352](https://www.fltk.org/str.php?L3352) "Tiny window problem if child group larger than window" in commit 74dd5164d3da59c99efc451438091c6a40327e0b introduced a regression as reported in fltk.coredev, thread ["Fl_GL_Window Resizing"](https://groups.google.com/g/fltkcoredev/c/EMm6R6J-Yb0).

This issue is only about the regression that windows that set a resizable() but don't call size_range() can no longer be resized smaller than their original window sizes.

**Comment by erco77** [comment]
I added comment #&#x2060;19 to the str, where the notes are relevant to the patches there:

> Albrecht, I think [this new issue] is what the bottom part of Comment #&#x2060;5 alludes to:
>
>    > v1 patch being path of least resistance
>    > v2 patch which _redefines_ the min size behavior
>
> Seems like the V2 path was taken (redefines resize behavior),
> but V1 is perhaps "better" if we want to continue the old behavior.

**Comment by Albrecht-S** [comment]
@erco77 wrote:
> V1 is perhaps "better" if we want to continue the old behavior.

Thanks, this is helpful. I took a quick look at the patch but I need to think about it. My brain is already half asleep and I'm going to watch SuperBowl which starts in my TZ around 11pm or midnight. ;-)

I'm not sure if we want to continue the old behavior but I'm sure the current behavior is not what we want. I noticed this myself in fluid where you can't resize the main window below its original size (same behavior as Rob reported) but I didn't get that it was a regression. I should have investigated...

Today I was trying to find a better way to calculate the "auto" size_range() values in a single function rather than in three different places in the code but I'll continue that tomorrow. My plan is:
1. find a good place for the new method, presumably in Fl_Window_Driver
2. put the new implementation (whatever it is) in this spot to avoid duplicated code.

**Comment by Albrecht-S** [comment]
Status report: Branch [issue-392_size_range](https://github.com/Albrecht-S/fltk/tree/issue-392_size_range) in my fork can be used to test the new functionality. Test/demo programs are included and built automatically with CMake (test/winsize*.cxx).

This branch has some active printf test/debug statements which will be removed before the final merge.

The new `size_range` calculation can be found in method [Fl_Window::default_size_range()](https://github.com/Albrecht-S/fltk/blob/9501e10f3cf63fb6fff44941993f560c6dfd7c08/src/Fl_Window.cxx#L661). It's pretty long with many comments but my tests seem to indicate that this got all possible cases.

Documentation updates are still missing.

**Comment by Albrecht-S** [comment]
The following text is my attempt to document the default `size_range()` calculation, i.e. if the window has a `resizable()` widget and `size_range()` has not been called. The full documentation can be found [in my fork](https://github.com/Albrecht-S/fltk/blob/f67cb8f964f25ddb9465e97cbf07d9cd31a26ca2/src/Fl_Window.cxx#L642).

**_Documentation of default size_range calculation:_**

... FLTK tries to figure out the window's size range from the
setting of the window's resizable() widget as follows and roughly in
the given order.

- If resizable() is NULL (this is the default) then the window cannot
  be resized and the resize border and max-size control will not be
  displayed for the window.

- If either dimension of resizable() is zero, then the window cannot
  resize in that direction.

- The resizable() widget is clipped to the window area.

- The non-resizable portion of the window is calculated as the difference
  of the window's size and the clipped resizable() widget's size.

- If either dimension of the clipped resizable() widget is greater
  than 100, then 100 is considered its minimum width/height. This
  allows the resizable widget to shrink below its original size.

- Finally the minimum width/height of the window is set to the
  non-resizable portion plus the width/height of the resizable()
  widget as calculated above.

In simple words:
- It is assumed that the resizable() widget can be indefinitely
  enlarged and/or shrunk to a minimum width/height of 100 unless
  it is smaller than that, which is then considered the minimum.
- The window's size_range() minimum values are set to the sum
  of the non-resizable portion of the window and the previously
  calculated minimum size of the resizable() widget.

Examples:
```
  Fl_Window win(400, 400);
  win.resizable(win);
  // win.size_range(100, 100, 0, 0);
```

The minimum size of the resizable is 100, hence the minimum size
of the total window is also 100 in both directions.

```
  Fl_Window win(400, 400);
  Fl_Box box(20, 20, 360, 360);
  win.resizable(box);
  // win.size_range(140, 140, 0, 0);
```

The calculated minimum width and height would be 20 + 100 + 20 in both
dimensions.

```
  Fl_Window win(400, 400);
  Fl_Box box(200, 0, 500, 300); // note: width 500 too large: clipped
  win.resizable(box);
  // win.size_range(300, 200, 0, 0);
```

The width of the resizable is clipped to 200, hence the minimum size of
the total window is also 200 (fix) + 100 (min. resizable) in x direction.
The minimum value in y direction is 100 (resizable) + 100 (fixed part).

The calculation is based on clipping the resizable widget to the window
area to prevent programming errors and the assumption that the resizable
widget can be shrunk to 100x100 or its original size, whichever is smaller.

If this is not what you want, please use Fl_Window::size_range()
explicitly so you can set any appropriate range.

**Comment by Albrecht-S** [comment]
The above documentation is part of my fix and documents the behavior as it was IMHO always intended but not implemented correctly. All comments and suggestions for improvements would be appreciated.

@erco77 The "magic 100" is useful (as documented) to make the window resizable by the user below its original size if the resizable is larger than 100 in any direction. This is the old behavior and should be kept. Other changes implement the previously documented behavior, the new docs try to clarify some details.

I tested this with several values including the "tiny window" example (STR 3352). I'll merge my branch (without the demo programs) if nobody objects during the next 36 hours. The [full diff](https://github.com/Albrecht-S/fltk/commit/982823a68fe51c3b07cfe65ffd08e6be6188f71b) can be found in my fork.

**Comment by wcout** [comment]
I tried your branch (Linux only). It works as described.

Though I don't know what how the following code *should* behave. It does not behave like before.
Previously it let the window resize in both directions and the group appears then. Now it will allow only vertical resize and the group does not appear:


```cxx
#include <FL/Fl_Window.H>
#include <FL/Fl_Group.H>
#include <FL/Fl.H>

int main()
{
	Fl_Window *win = new Fl_Window(100, 100);
	Fl_Group *group = new Fl_Group(-100, -100, 50, 50);
	group->box(FL_FLAT_BOX);
	group->color(FL_GREEN);
	win->resizable(group);
	win->end();
	win->show();
	return Fl::run();
}
```

**Comment by Albrecht-S** [comment]
Again, you wrote: "I don't know what how the following code should behave".

I did not directly answer this question (if it was one). The intention is to let the window resize in both directions, i.e. the same behavior in x direction as it has in y direction.

**Comment by wcout** [comment]
> You wrote: "I don't know what how the following code should behave".

I probably meant, if a resizable is outside of the window, should the window then be resizable at all. Didn't know if the bug was "one direction still resizable" or "one direction not resizable".

But that's all good now and consistent too.

**Comment by Albrecht-S** [comment]
Fixed in commit 091712bea8ff5aab89f0f8483ca572c118ca5715, documentation modified accordingly.

For a summary and reasoning of the implemented changes see also [STR 3352](https://www.fltk.org/str.php?L3352), comment 20.

**Comment by wcout** [comment]
Sorry to bother, but I'm observing some issues after this fix (Linux/Ubuntu 20.04) in  https://github.com/fltk/fltk/commit/091712bea8ff5aab89f0f8483ca572c118ca5715

- Ctrl +/- does not work correctly - the windows are not resized with the content
- glpuzzle opens 2 very tiny windows

I went back to the commit before and all is well again. No I see this behaviour already present in the Branch [issue-392_size_range](https://github.com/Albrecht-S/fltk/tree/issue-392_size_range) in the @Albrecht-S fork.

**Comment by Albrecht-S** [comment]
@wcout Thanks for bothering ;-)

I confirm both bugs and I'm reopening the issue.

**Comment by Albrecht-S** [comment]
Both bugs have been fixed in commit 5ebb76050b77aaf6047669968fc3919f3dd043f5.
Branch `issue-392_size_range` has been deleted from my fork.
